core/src/inputs/advice/merkle_path_set.rs

```diff
@@ -0,0 +1,339 @@
+        }
+        if depth != self.total_depth {
+            return Err(AdviceSetError::InvalidDepth(self.total_depth, depth));
+        }
+
+        let pos = 2u64.pow(depth) + index;
+        let index = (pos / 2) as u64;
+
+        match self.paths.get(&index) {
+            None => Err(AdviceSetError::NodeNotInSet(index)),
+            Some(path) => {
+                if is_even(pos) {
+                    Ok(path[0])
+                } else {
+                    Ok(path[1])
+                }
+            }
+        }
+    }
+
+    /// Returns a Merkle path to the node at the specified index. The node itself is
+    /// not included in the path.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified index not valid for the depth of structure.
+    /// * Node of the requested path does not exist in the set.
+    pub fn get_path(&self, depth: u32, index: u64) -> Result<Vec<Word>, AdviceSetError> {
+        if index >= 2u64.pow(self.total_depth) {
+            return Err(AdviceSetError::InvalidIndex(self.total_depth, index));
+        }
+        if depth != self.total_depth {
+            return Err(AdviceSetError::InvalidDepth(self.total_depth, depth));
+        }
+
+        let pos = 2u64.pow(depth) + index;
+        let index = pos / 2;
+
+        match self.paths.get(&index) {
+            None => Err(AdviceSetError::NodeNotInSet(index)),
+            Some(path) => {
+                let mut local_path = path.clone();
+                if is_even(pos) {
+                    local_path.remove(0);
+                    Ok(local_path)
+                } else {
+                    local_path.remove(1);
+                    Ok(local_path)
+                }
+            }
+        }
+    }
+
+    /// Replaces the leaf at the specified index with the provided value.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * Requested node does not exist in the set.
+    pub fn update_leaf(&mut self, index: u64, value: Word) -> Result<(), AdviceSetError> {
+        let depth = self.depth();
+        if index >= 2u64.pow(depth) {
+            return Err(AdviceSetError::InvalidIndex(depth, index));
+        }
+        let pos = 2u64.pow(depth) + index;
+
+        let path = match self.paths.get(&(pos / 2)) {
+            None => return Err(AdviceSetError::NodeNotInSet(index)),
+            Some(path) => path.clone(),
+        };
+
+        // Fill old_hashes vector -----------------------------------------------------------------
+        let (old_hashes, _) = compute_path_trace(&path, depth, index);
+
+        // Fill new_hashes vector -----------------------------------------------------------------
+        let pos = 2u64.pow(depth) + index;
+
+        if is_even(pos) {
+            self.paths.get_mut(&(pos / 2)).unwrap()[0] = value;
+        } else {
+            self.paths.get_mut(&(pos / 2)).unwrap()[1] = value;
+        }
+
+        let path = self.paths.get(&(pos / 2)).unwrap().clone();
```

Comment:
sorry - just noticed this: is there a reason we can't use `path` retrieved on line 166?

Reply:
We can't use it because the path we are working with has been changed in lines 177-181 --- the hash of the leaf has been updated. So we have to update our `path` by getting it again

Reply:
I think you should be able to get a mutable reference to the path to start with, and then you won't need to get path from the set again. Maybe something like this:
```Rust
let path = match self.paths.get_mut(&(pos / 2)) {
    None => return Err(AdviceSetError::NodeNotInSet(index)),
    Some(path) => path,
};

// Fill old_hashes vector
let (old_hashes, _) = compute_path_trace(&path, depth, index);

// Fill new_hashes vector
let pos = 2u64.pow(depth) + index;

if is_even(pos) {
    path[0] = value;
} else {
    path[1] = value;
}

let (new_hashes, new_root) = compute_path_trace(&path, depth, index);
self.root = new_root;
```

Reply:
Indeed! Somehow I didn't think about it. Thank you!